### PR TITLE
Issues/1350

### PR DESCRIPTION
**Pull request by cixil**
Not sure why the pipfile was included

Added 5 tests
- Create network source with exclude host list - PASS
-  Create satellite source with exclude host list - FAIL
-  Create vcenter source with exclude host list - FAIL
-  Update satellite source with exclude host list - FAIL
-  Update vcenter source with exclude host list - FAIL

### CODE REVIEW

File: quipucords/api/source/tests_source.py [hunk 1]
```diff
@@ -86,13 +86,17 @@ def create_expect_400(self, data, expected_response=None):
         """We will do a lot of create tests that expect HTTP 400s."""
         response = self.create(data)
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
+        if response.status_code != 400:
+            print(response.json())
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: Is this debug code?

File: quipucords/api/source/tests_source.py [hunk 2]
```diff
@@ -86,13 +86,17 @@ def create_expect_400(self, data, expected_response=None):
...
         if expected_response:
             response_json = response.json()
             self.assertEqual(response_json, expected_response)
 
     def create_expect_201(self, data):
         """Create a source, return the response as a dict."""
         response = self.create(data)
+        if response.status_code != 201:
+            print(response.json())
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: Debug?

File: quipucords/api/source/tests_source.py [hunk 3]
```diff
@@ -937,6 +972,27 @@ def test_update_vc_more_than_one_host(self):
                                    format='json')
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
 
+    def test_update_vc_with_exclude_host(self):
+        """VC - Fail more than one host."""
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: This docstring should say something more along the lines of """VC - Fail with exclude hosts."""

File: quipucords/api/source/tests_source.py [hunk 4]
```diff
@@ -1035,16 +1091,25 @@ def test_create_sat_with_hosts(self):
...
+        self.create_expect_400(
+            {'name': 'source1',
+             'source_type': Source.SATELLITE_SOURCE_TYPE,
+             'hosts': ['1.2.3.4'],
+             'exclude_hosts': ['1.2.3.4'],
+             'credentials': [self.sat_cred_for_upload]})
+
     def test_create_sat_with_host_range(self):
-        """A vcenter source must not have multiple hosts."""
+        """A satellite source must not have multiple hosts."""
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: Nice catch!

File: quipucords/api/source/tests_source.py [hunk 5]
```diff
@@ -1035,16 +1091,25 @@ def test_create_sat_with_hosts(self):
...
         self.create_expect_400(
             {'name': 'source1',
              'source_type': Source.SATELLITE_SOURCE_TYPE,
              'hosts': ['1.2.3.4/5'],
              'credentials': [self.sat_cred_for_upload]})
 
     def test_create_req_type(self):
-        """A vcenter source must have an type."""
+        """A satellite source must have an type."""
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: I think that this one would actually be a vcenter source, since the cred being used is vcenter.

File: quipucords/api/source/tests_source.py [hunk 6]
```diff
@@ -1116,6 +1181,27 @@ def test_update_sat_more_than_one_hosts(self):
                                    format='json')
         self.assertEqual(response.status_code, status.HTTP_400_BAD_REQUEST)
 
+    def test_update_sat_more_exclude_hosts(self):
+        """Sat- Fail update due to multiple hosts."""
```
Comment (abaiken) [inline quipucords/api/source/tests_source.py]: This docstring should be like the vcenter one : """Sat - Fail update due to exclude hosts."""

**Comment by kholdaway** [comment]
Add files like these to your global .gitignore (remove them from the PR).
- quipucords/api/source/.model.py.swp

File: quipucords/api/source/tests_source.py (hunk 1 above)
Comment (cixil) [reply quipucords/api/source/tests_source.py]: yes, will remove

File: quipucords/api/source/tests_source.py (hunk 5 above)
Comment (cixil) [reply quipucords/api/source/tests_source.py]: It was among all the other satellite tests, I'll move it to the vcenter section

File: quipucords/api/source/serializer.py [hunk 7]
```diff
@@ -197,6 +210,7 @@ def create(self, validated_data):
             source.options = options
 
         source.hosts = json.dumps(hosts_list)
+        source.exclude_hosts = json.dumps(exclude_hosts_list)
```
Comment (abaiken) [inline quipucords/api/source/serializer.py]: Add a check to make sure exclude_hosts isn't None before we assign it to the source

File: quipucords/api/source/view.py [hunk 8]
```diff
@@ -175,6 +175,7 @@ def create(self, request, *args, **kwargs):
                 if convert_to_boolean(scan):
                     # Grab the source id
                     source_id = response.data['id']
+
```
[the remark this answers is not in the document]
Comment (cixil) [reply quipucords/api/source/view.py]: It makes it prettier

[the remark this answers is not in the document]
Comment (cixil) [reply quipucords/api/source/view.py]: Before it looked like this
```
176                     # Grab the source id                                                                                                                                       
177                     source_id = response.data['id']                                                                                                                            
178                     # Create the scan job                                                                                                                                      
179                     scan_job = ScanJob(scan_type=ScanTask.SCAN_TYPE_CONNECT)                                                                                                   
180                     scan_job.save()                                                                                                                                            
181                                                                                                                                                                                
182                     # Add the source                                                                                                                                           
183                     scan_job.sources.add(source_id)                                                                                                                            
184                     scan_job.save()                                                                                                                                            
185                     json_source['most_recent_connect_scan'] = scan_job.id                                                                                                      
186                                                                                                                                                                                
187                     # Start the scan                                                                                                                                           
188                     start_scan.send(sender=self.__class__, instance=scan_job)       
```

File: .gitignore [hunk 9]
```diff
@@ -76,6 +76,9 @@ target/
 # pyenv
 .python-version
 
+# Pipfile
```
Comment (kholdaway) [inline .gitignore]: Why is this here?

File: quipucords/api/source/serializer.py [hunk 10]
```diff
@@ -282,6 +308,10 @@ def update(self, instance, validated_data):
             hosts_data = json.dumps(hosts_list)
             instance.hosts = hosts_data
 
+        if exclude_hosts_list:
+            exclude_hosts_data = json.dumps(exclude_hosts_list)
```
Comment (kholdaway) [inline quipucords/api/source/serializer.py]: Just remove the temp variable from these two lines

File: quipucords/api/source/serializer.py [hunk 11]
```diff
@@ -349,16 +379,16 @@ def validate_name(name):
...
     # pylint: disable=too-many-locals, too-many-branches, too-many-statements
     @staticmethod
-    def validate_hosts(hosts):
-        """Make sure the hosts list is present."""
-        hosts_list = json.loads(hosts)
-        if not isinstance(hosts_list, list):
+    def validate_ipaddr_list(hosts):
+        """Make sure the hosts list is present and has valid IP addresses."""
+        ipaddr_list = json.loads(hosts)
+        if not isinstance(ipaddr_list, list):
```
Comment (kholdaway) [inline quipucords/api/source/serializer.py]: This error message should be generalized a bit.  It indicates the problem is with host.

File: quipucords/api/source/view.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (kholdaway) [reply quipucords/api/source/view.py]: 👍

**Comment by kholdaway** [comment]
Added one more change (small).  After builds pass feel free to merge.
